doc/design/scheduler.md

```diff
@@ -0,0 +1,177 @@
+# Scheduler for TrainingJob
+
+## Background
+
+We are going to define PaddlePaddle cluster job as a Kubernetes
+[TPR](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-third-party-resource/) or
+[CRD](https://kubernetes.io/docs/concepts/api-extension/custom-resources/).
+Each job is described using a `yaml` representation called
+[TrainingJob](../autoscale/README.md). When a `TrainingJob` resource is
+submitted to Kubernetes cluster, our customized `controller` program will
+receive an event informing the resource creation/deletion.
+
+The `controller` program should contain the following core functions:
+
+- Parser to parse `TrainingJob` resource to corresponding job components,
+  including:
+    - `ReplicaSet` of master process
```

Comment:
Using `ReplicaSet` / `StatefulSet` / `Job` means we will depend on Kubernetes' scheduler for scheduling Pods. Since we are creating our own scheduler, should we rely on Kubernetes' scheduler or not? What is the pros and cons of both cases?

Reply:
Yes, you are right. Not using default k8s scheduler will let us have more control over `TraningJobs` see [here](https://kubernetes.io/docs/tasks/administer-cluster/configure-multiple-schedulers/). The scheduler is in charge of putting pods on nodes.

Pros:
- Can add weight to resource types like GPU when scheduling pods.
- Scaling and scheduling can be in the same process.
- New resource types can be taken care of like FPGA.

Cons:
- The core function of schedule pods to nodes seems to be same for the `TrainginJob` scheduler. Resource request per node won't be changed, we only change the number of pods to run, which is already done by `autoscaler`.
- Hard to implement, we have to implement leader selection for "High Availability".

I think using `default-scheduler` for the pod->nodes job is enough currently, we only need to queue `TrainingJobs` by priority and hand them to k8s, for now.

Reply:
I see, thanks. Agree that using the default scheduler is better for our current use case.

Another related question: do we need to use k8s Job / StateSet, or maybe we can control the Pods directly (but still using the default scheduler).

Reply:
> Another related question: do we need to use k8s Job / StateSet, another possibility is we can submit the creation and deletion of Pods directly (but still using the default scheduler).

This is possible and may be useful. Instead, the controller has to track all pods' status, which is implemented in the k8s Job/StateSet controller.

Pros for directly control pods:
- control scale by pod status, i.e. scale up/down the slowest pod (pod runs the smallest batch-id)
- dynamically change resource request of pods.

Reply:
@helinwang @Yancey1989 what you think of "submit the creation and deletion of Pods directly", I'll update the doc if we all agree with this.

Reply:
> submit the creation and deletion of Pods directly

That's a great idea and agree with this. But maybe we can implement the scheduler firstly?

Reply:
@gongweibao : please take a look at this discussion, it's related to your converting Python start job to Go controller.

Reply:
OK!